docs/generating-a-provider.md

```diff
@@ -0,0 +1,391 @@
+       -       tf "github.com/hashicorp/terraform-provider-hashicups/hashicups"
+       +       tf "github.com/turkenh/terraform-provider-github/v4/github"
+       )
+  
+       const resourcePrefix = "github"
+       ```
+
+       Run:
+       ```bash
+       go mod tidy
+       ```
+
+       Please note, we are temporarily using a [fork] of
+       [terraform-provider-github] repo as a workaround to [this issue].
+
+    4. If your provider uses an old version (<v2) of [terraform-plugin-sdk],
+       convert resource map to v2 schema as follows (uncomment related section):
+
+       ```go
+       import (
+           "github.com/crossplane-contrib/terrajet/pkg/types/conversion"
+       )
+       
+       resourceMap := conversion.GetV2ResourceMap(tf.Provider())
+       ```
+
+       And in `go.mod` file, set the following `replace directive`
+       (uncomment related section):
+
+       ```
+       github.com/hashicorp/terraform-plugin-sdk => github.com/turkenh/terraform-plugin-sdk v1.17.2-patch1
+       ```
+
+       Run:
+       ```bash
+       go mod tidy
+       ```
+
+5. Implement `ProviderConfig` logic. In `provider-tf-template`, there is already
+   a boilerplate code in file `internal/clients/${ProviderNameLower}.go` which
+   takes care of properly fetching secret data referenced from `ProviderConfig`
+   resource.
+
+   For our GitHub provider, we need to check [Terraform documentation for provider
+   configuration] and provide the keys there:
+
+   ```go
+   const (
+     keyBaseURL = "base_url"
+     keyOwner = "owner"
+     keyToken = "token"
+
+     // GitHub credentials environment variable names
+     envToken = "GITHUB_TOKEN"
+   )
+
+   func TerraformSetupBuilder(version, providerSource, providerVersion string) terraform.SetupFn {
+     ...
+     // set provider configuration
+     ps.Configuration = map[string]interface{}{}
+     if v, ok := githubCreds[keyBaseURL]; ok {
+         ps.Configuration[keyBaseURL] = v
+     }
+     if v, ok := githubCreds[keyOwner]; ok {
+         ps.Configuration[keyOwner] = v
+     }
+     // set environment variables for sensitive provider configuration
+     ps.Env = []string{
+         fmt.Sprintf(fmtEnvVar, envToken, githubCreds[keyToken]),
+     }
+     return ps, nil
+   }
+   ```
+
+6. Before generating all resources that the provider has, let's go step by step
+   and only start with generating `github_repository` and `github_branch`
+   resources.
+
+   To limit the resources to be generated, we need to provide an include list
+   option with `tjconfig.WithIncludeList` in file `config/provider.go`:
+
+   ```go
+   pc := tjconfig.NewProvider(resourceMap, resourcePrefix, modulePath,
+       tjconfig.WithDefaultResourceFn(defaultResourceFn),
+       tjconfig.WithIncludeList([]string{
+           "github_repository$",
+           "github_branch$",
+       }))
+   ```
+
+7. Finally, we would need to add some custom configurations for these two
```

Comment:
Because the custom config doc goes into great detail, what do you think leaving provider generation part end here? i.e. use `.*` to generate all and then let them know they can now provide custom configs by following that other guide.

Reply:
I was about to leave the same comment.

I'm about to run through this guide for a second provider and I plan on skipping steps 4, 6, and 7.  I'll use `.*`.

Once I have all of the CRDs generated I will revisit step 7 to set up references. I'll revisit step 6 if I have to remove any resources that don't convert well.

Reply:
I would prefer to create an end-to-end experience here. 
Starting from scratch and having some resources provisioned. I am a bit hesitant that if people try to generate all resources at this point, they might be discouraged due to some unexpected issues (e.g. conflicts due to a missing resource configuration).

I've extended the note after step 8 to mention how all resources could be generated and configuring them could be a next step.